File: src/main/java/chloe/movietalk/auth/JwtProvider.java
package chloe.movietalk.auth;

import chloe.movietalk.dto.response.user.UserInfo;
import io.jsonwebtoken.*;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import io.jsonwebtoken.security.SecurityException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.security.Key;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

@Slf4j
@Component
public class JwtProvider {

    private final Key secretKey;

    private final Long accessTokenExpTime;

    private final Long refreshTokenExpTime;

    public JwtProvider(
            @Value("${jwt.secret}") final String secretKey,
            @Value("${jwt.access_expiration_time}") final Long accessTokenExpTime,
            @Value("${jwt.refresh_expiration_time}") final Long refreshTokenExpTime) {
        byte[] keyBytes = Decoders.BASE64.decode(secretKey);
        this.secretKey = Keys.hmacShaKeyFor(keyBytes);
        this.accessTokenExpTime = accessTokenExpTime;
        this.refreshTokenExpTime = refreshTokenExpTime;
    }

    public String generateAccessToken(UserInfo user) {
        Long now = System.currentTimeMillis();
        return Jwts.builder()
                .setHeader(createHeader())
                .setClaims(createClaims(user))
                .setSubject(user.getId().toString())
                .setIssuedAt(new Date(now))
                .setExpiration(new Date(now + accessTokenExpTime))
                .signWith(secretKey, SignatureAlgorithm.HS256)
                .compact();
    }

    public String generateRefreshToken(UUID userId) {
        Long now = System.currentTimeMillis();
        return Jwts.builder()
                .setHeader(createHeader())
                .setSubject(userId.toString())
                .setIssuedAt(new Date(now))
                .setExpiration(new Date(now + refreshTokenExpTime))
                .signWith(secretKey, SignatureAlgorithm.HS256)
                .compact();
    }

    public UUID getUserId(String token) {
        return UUID.fromString(parseClaims(token).getSubject());
    }

    public Date getExpiration(String token) {
        return parseClaims(token).getExpiration();
    }

    public Claims parseClaims(String token) {
        return Jwts.parserBuilder()
                .setSigningKey(secretKey)
                .build()
                .parseClaimsJws(token)
                .getBody();
    }

    public boolean isTokenExpired(String token) {
        try {
            Date expirationDate = parseClaims(token).getExpiration();
            return expirationDate != null && expirationDate.before(new Date());
        } catch (ExpiredJwtException e) {
            return true;
        }
    }

    public boolean isValidToken(String token) {
        try {
            Jwts.parserBuilder()
                    .setSigningKey(secretKey)
                    .build()
                    .parseClaimsJws(token);
            return true;
        } catch (SecurityException e) {
            log.warn("Invalid JWT signature", e);
        } catch (MalformedJwtException e) {
            log.warn("Invalid JWT token", e);
        } catch (ExpiredJwtException e) {
            log.warn("Expired JWT", e);
        } catch (UnsupportedJwtException e) {
            log.warn("Unsupported JWT", e);
        } catch (IllegalArgumentException e) {
            log.warn("JWT claims string is empty", e);
        }

        return false;
    }

    private Map<String, Object> createHeader() {
        Map<String, Object> header = new HashMap<>();
        header.put("typ", "JWT");
        header.put("alg", "HS256");
        return header;
    }

    private Map<String, Object> createClaims(UserInfo user) {
        Map<String, Object> claims = new HashMap<>();
        claims.put("userId", user.getId());
        claims.put("email", user.getEmail());
        claims.put("userRole", user.getRole());
        return claims;
    }
}

File: src/main/java/chloe/movietalk/auth/JwtProvider.kt
package chloe.movietalk.auth

import chloe.movietalk.dto.response.user.UserInfo
import io.jsonwebtoken.*
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import io.jsonwebtoken.security.SecurityException
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.security.Key
import java.util.*

@Component
class JwtProvider(
    @Value("\${jwt.secret}") secretKey: String,
    @Value("\${jwt.access_expiration_time}") accessTokenExpTime: Long,
    @Value("\${jwt.refresh_expiration_time}") refreshTokenExpTime: Long
) {
    private val secretKey: Key
    private val accessTokenExpTime: Long
    private val refreshTokenExpTime: Long

    init {
        val keyBytes = Decoders.BASE64.decode(secretKey)
        this.secretKey = Keys.hmacShaKeyFor(keyBytes)
        this.accessTokenExpTime = accessTokenExpTime
        this.refreshTokenExpTime = refreshTokenExpTime
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(JwtProvider::class.java)
    }

    fun generateAccessToken(user: UserInfo): String {
        val now = System.currentTimeMillis()
        return Jwts.builder().apply {
            setHeader(createHeader())
            setClaims(createClaims(user))
            setSubject(user.id.toString())
            setIssuedAt(Date(now))
            setExpiration(Date(now + accessTokenExpTime))
            signWith(secretKey, SignatureAlgorithm.HS256)
        }.compact()
    }

    fun generateRefreshToken(userId: UUID): String {
        val now = System.currentTimeMillis()
        return Jwts.builder().apply {
            setHeader(createHeader())
            setSubject(userId.toString())
            setIssuedAt(Date(now))
            setExpiration(Date(now + refreshTokenExpTime))
            signWith(secretKey, SignatureAlgorithm.HS256)
        }.compact()
    }

    fun getUserId(token: String): UUID {
        return UUID.fromString(parseClaims(token).subject)
    }

    fun getExpiration(token: String): Date {
        return parseClaims(token).expiration
    }

    fun parseClaims(token: String): Claims {
        return Jwts.parserBuilder()
            .setSigningKey(secretKey)
            .build()
            .parseClaimsJws(token)
            .body
    }

    fun isTokenExpired(token: String): Boolean {
        try {
            return parseClaims(token).expiration?.before(Date())
                ?: return true
        } catch (e: ExpiredJwtException) {
            return true
        }
    }

    fun isValidToken(token: String): Boolean {
        return try {
            Jwts.parserBuilder()
                .setSigningKey(secretKey)
                .build()
                .parseClaimsJws(token)
            true
        } catch (e: SecurityException) {
            log.warn("Invalid JWT signature", e)
            false
        } catch (e: MalformedJwtException) {
            log.warn("Invalid JWT token", e)
            false
        } catch (e: ExpiredJwtException) {
            log.warn("Expired JWT", e)
            false
        } catch (e: UnsupportedJwtException) {
            log.warn("Unsupported JWT", e)
            false
        } catch (e: IllegalArgumentException) {
            log.warn("JWT claims string is empty", e)
            false
        }
    }

    private fun createHeader(): MutableMap<String, Any> {
        val header: MutableMap<String, Any> = HashMap<String, Any>()
        header.put("typ", "JWT")
        header.put("alg", "HS256")
        return header
    }

    private fun createClaims(user: UserInfo): MutableMap<String, Any> {
        val claims: MutableMap<String, Any> = HashMap<String, Any>()
        claims.put("userId", user.id)
        claims.put("email", user.email)
        claims.put("userRole", user.role)
        return claims
    }
}
